### ISSUE TITLE
Consecutively  Service task scheduling mode - scale a task when all other tasks report healthy

### ISSUE THREAD

**Issue by krasi-georgiev** [moby/moby#30194]
I am working on a galera cluster using percona and when you scale up any new node need to connect to a running node to sync with the cluster

currently what happens is that all tasks start at the same time and the sync is messed up and most containers will shutdown.

something like docker service create --consecutively
start a new task only when all other tasks report halthy

**Comment by fntlnz** [comment]
Have you tried to use the healthcheck feature and act based on healthchecks?

**Comment by krasi-georgiev** [comment]
let me know if  you think I can implement it in the following scenario.

https://github.com/vipconsult/percona-docker/blob/master/pxc-57/entrypoint.sh

when the task starts it checks if the service has any other running tasks and connects to one of the running tasks to sync with the mysql cluster

this is a code form the entrypoint.sh
readarray -t serviceIpArray <<< "$(dig -t A +short tasks.$SERVICE_NAME)"

if serviceIpArray is empty that means that this is the first task so it needs to bootstrap a new cluster , for any additional starting task  serviceIpArray should not be empty so instead of boostrapping a new cluster should connect to the existing tasks to sync.

currently when the tasks all start at the same time serviceIpArray is empty for all tasks so each task bootstraps a new cluster.

I can't think of any better way to design this.

it would be interesting to see any other better design ideas.

**Comment by krasi-georgiev** [comment]
the only way I can see this working is if the scheduler doesn't run another task until the pervious task is in healthy this way any new task will know that it needs to join an existing cluster.

**Comment by krasi-georgiev** [comment]
@fntlnz - can you give an example

**Comment by fntlnz** [comment]
To give you an opinion I have to know how you are starting all this, you are talking about tasks so I suppose that you're using swarmkit. Also, what sets the `SERVICE_NAME` env variable?

Services are mostly inteneded to replicate the same service, if I understand correctly you're using them to create a mysql replica  which is not composed by three instances of the same thing but is three different things that are synced at application level.

A good solution might be having a stack in which you deploy three different mysql services, in this way you can also take benefit of the service discovery because you'll have three different names, one of each service. This enables you also to use healthchecks to wait for the first one to come up and then start the others.

**Comment by krasi-georgiev** [comment]
SERVICE_NAME is set as an -e SERVICE_NAME= when creating the service

> Services are mostly inteneded to replicate the same service, if I understand correctly you're using them to create a mysql replica which is not composed by three instances of the same thing but is three different things that are synced at application level.
**correct**

> A good solution might be having a stack in which you deploy three different mysql services, in this way you can also take benefit of the service discovery because you'll have three different names, one of each service. This enables you also to use healthchecks to wait for the first one to come up and then start the others.
**this solution defeats the whole purpose of a service and scaling.**

**Comment by fntlnz** [comment]
>> this solution defeats the whole purpose of a service and scaling.
@krasi-georgiev sorry, I misunderstood your points

You are on the right lane with this command `readarray -t serviceIpArray <<< "$(dig -t A +short tasks.$SERVICE_NAME)"`

I would just add an healthcheck which based on your commands waits for the N services to be listed and then if I am the one with the lower IP I do the boostrap, instead if i'm one of the others I can try to connect to the first one to see if it's ready (supposing that there's is a way to know this information remotely)



Wdyt?

**Comment by fntlnz** [comment]
@krasi-georgiev I will try to write down an example based on your image, just wait a couple of hours :P

**Comment by chanwit** [comment]
@krasi-georgiev I'm running pxc in a pre-production environment (thanks for the great Percona's image) and I'm finding that this is a specific behaviour to Galera.

I too would like to see this kind of option to be implemented in SwarmKit.

**Comment by krasi-georgiev** [comment]
@fntlnz yes that would work , but how would the first node know that no other nodes still in healthy(starting) so it is ok to bootstrap a new cluster. The specific problem is that at any given moment there might be more than one node with health (starting) so dig -t A +short tasks.$SERVICE_NAME will be empty and all these nodes will bootstrap a new cluster given the result that dig returns no other IP

@chanwit  thanks , it is crucial to find a solution to this specific problem and it would be amazingly easy to scale the percona cluster.

**Comment by krasi-georgiev** [comment]
@fntlnz I would be very interested to read your idea.

**Comment by krasi-georgiev** [comment]
@thaJeztah 
do you think if I try to implement this it will get merged ?
could you maybe discuss it with the devs. 

I can't think of any other way to properly scale mysql,galera,percona,mariadb without this option.

**Comment by fntlnz** [comment]
@krasi-georgiev sorry for the late response, I was talking with @thaJeztah to better understand the use cases and a possible design of this kind of feature.

The best thing we can do now (before than actually implementing)  is to try to clarify the design and possible missing parts in order to consider working and putting effort on this.

Specifically, with @thaJeztah we were thinking about a design similar to how swarm actually works where the first node is "special" because it creates token for the other managers, in the same way we can have something like a "bootstrap" task which actually setup all the things.

In any case, for a mysql cluster for example, the bootstrap node could create a master-slave setup where the slave is not exactly the same thing as the master, you will likely read from it but not write, so again we face a possible trouble.

**Comment by krasi-georgiev** [comment]
great , lets discuss it then :)

with percona and mariadb(I think it is the same with any mysql cluster) it works a bit different than your description or maybe I didn't understand what you mean by "bootstrap task"

the cluster works on a master master setup - when you read you read from one node but when you write you write to all nodes

but the main problem as you mentioned is the initial bootstrap. 

In order to create a cluster all nodes need to know if there is another healthy node to connect to get the data from.  

each node can be a donor to only a single node - meaning if more than one node try to connect to the same donor one transfer will fail.

so in the current setup when all nodes start at the same time I can't think of any way to properly order the start up sequence. unless I start the containers one by one. scale=1  then scale=2 and so on.


so this is what happens:

first service task:
readarray -t runningTasks <<< "$(dig -t A +short tasks.mysql)"
runningTasks is empty so the container bootstraps a new cluster

second service task
readarray -t runningTasks <<< "$(dig -t A +short tasks.mysql)"
runningTasks include the ip of the first task so the cluster is given an exisitng ip and it connect to it to get the data and it joins the existing cluster

third service task
readarray -t runningTasks <<< "$(dig -t A +short tasks.mysql)"
runningTasks includes the ip of the first and the second task so the cluster is given 2 ips to  connect to get the data.(if one node is busy as a donor it will automatically use the second one.)

.
.
.
n-th service container
readarray -t runningTasks <<< "$(dig -t A +short tasks.mysql)"
runningTasks includes the ip of all healthy  task so the cluster is given many ips to  connect to get the data.

any new node will fail to join if there are no nodes that can be selected as a donor 
(ie all nodes are currently donors donors)


If we manage to solve this problem scaling would be a very pleasant experience :)

**Comment by krasi-georgiev** [comment]
docker service create --schedule=serial
docker service create --schedule=parallel (default)

this is how I see it

**Comment by krasi-georgiev** [comment]
ping @fntlnz

**Comment by fntlnz** [comment]
I was saying:
If you could have the ability to define a sort of **boostrap** task which is guaranteed to start before the others you could do all your setup there and then "release" the creation of the other tasks.

Like:

```bash
$ docker service create \
--name=mydbwhichscales --replicas=3 \
--bootstrap=mybootstrap.sh me/percona:5.7 
```
This would create a task, which will execute the `mybootstrap.sh` script, wait for it to complete and then release the other 2 tasks which can then connect to the first task to replicate.

We could define some convention for bootstrap  scripts like it now exists for healthchecks, like, If it exits with `0` the bootstrap is completed, if exits with `1` there has been an error


We could even have a directive in the Dockerfile to allow defining bootstrap tasks at an image creation level.


Last but not least: The bootstrap lifecycle is executed only at service creation.


Wdyt?

**Comment by krasi-georgiev** [comment]
the bootstrap is different depending on whether it is the first node or an additional one

not sure how would you know if this is the first task or an additional task ?

**Comment by fntlnz** [comment]
The bootstrap task is executed only on the first node and prior to the entrypoint.

The flow could be something like:

```
                             service created w/bootstrap
                                         +
                                         |
                                         |
                                         v
                        setup all the networking, volumes, etc..
                                         +
                                         |
                                         v
        +--------------------> schedule the first task
      retry                              +
        ^                                v
        |            execute the bootstrap script on the first task
        |                            exit with
        |                            1       0
        +----------------------------+       +
                                             |
                                        +----+
                                        |
                                        v
                             schedule the remaining tasks
```

Then the retry could have a max num of attempts or something like that.

It's more clear?

**Comment by runcom** [comment]
> We could even have a directive in the Dockerfile to allow defining bootstrap tasks at an image creation level.

I don't think the image (and Dockerfile) is the right level to put this in _at all_.

**Comment by krasi-georgiev** [comment]
this design might work , but looks a bit overcomplicated.

one possible issue as that after you schedule all additional tasks they will all try to connect to the boostrapped node and only one will connect as the bootstraped node can be a donor to a single node.
So what will happen is that if you schedule 50 replicas 1 will bootstrap , 1 will connect to the donor and all other 48 will connect and fail to sync so potentially the whole service will fail when it reaches the retry count.


I would love to hear some more scenarios with pro and cons about both design paths.


on a side note , which design path would be easier to implement ?

docker service create --schedule=serial
docker service create --schedule=parallel (default)

**Comment by fntlnz** [comment]
@runcom i agree, I thought that could be the same as health check but this is something happening at an higher level (e.g: it's only relevant for a task and not for a container)

**Comment by krasi-georgiev** [comment]
@fntlnz did you read my post , there is still an issue with the design you proposed.

**Comment by chanwit** [comment]
@krasi-georgiev I have implemented the consecutive algorithm and embedded into Docker (SwarmKit) 1.12  successfully, but I found that it's really hard to do so with 1.13 or later versions.

**Comment by thaJeztah** [comment]
@krasi-georgiev thanks for adding the additional information on how the process works (https://github.com/docker/docker/issues/30194#issuecomment-276220176); it's starting to make more sense now. Silly question; is the "1 node can only be donor for 1 replica" only during bootstrap, or is it during the entire lifetime? i.e. When _bootstrapping_; the first node is started, has data, a second node is added, copies the data from its donor, then the next node, etc. Once the nodes are up, there's no longer a "chain"?

**Comment by krasi-georgiev** [comment]
> 1 node can only be donor for 1 replica" only during bootstrap, or is it during the entire lifetime?

during the entire lifetime.
each healthy replica can be a donor only to a single node at any given time , that is why when you start a new replica you give it an array of all available healthy nodes so it can select a NON busy replica to sync from.

@vadimtk can correct me if I am wrong, but based on my tests this is how it behaves. 



**an example workflow that would work** 

service create mysqlCluster
-> first node started 
-> entrypoint.sh runs dig -t A +short tasks.mysqlCluster 
-> zero nodes returned so mysql bootstraps a new cluster : mysql --wsrep_cluster_address=""
-> waits to  be reported as healthy 
-> second node is started 
-> entrypoint.sh runs dig -t A +short tasks.mysqlCluster which returns 1 healthy node on "10.0.0.1"
-> mysql --wsrep_cluster_address="10.0.0.1"
-> waits to  be reported as healthy - which  is only after the sync with the donor is finished
-> another node is started
-> entrypoint.sh runs dig -t A +short tasks.mysqlCluster which returns 2 healthy nodes on "10.0.0.1,10.0.0.2"
-> mysql --wsrep_cluster_address="10.0.0.1,10.0.0.2"

at this stage you can even start 2 nodes as you have 2 available nodes that can serve as donors.

**Comment by vadimtk** [comment]
@krasi-georgiev you are correct.
When a node in the cluster becomes "DONOR" to a joining node, it can't handle other joiners.
But as soon as it finished with being "DONOR" and become "SYNC", it can server new joiners.

One can check the node's status by checking status variable "wsrep_local_state_comment", if this is not "Synced", then node can't serve as "DONOR"

**Comment by vadimtk** [comment]
@krasi-georgiev so as a workaround to the original problem you can check in loop `wsrep_local_state_comment` for existing tasks in `entrypoint.sh`  and continue the task only when you find a node with `wsrep_local_state_comment=Synced`. And some protection from the race condition would be useful (that's why I rely on DISCOVERY SERVICE, as it provides atomic operations)

**Comment by krasi-georgiev** [comment]
@vadimtk 

>  so as a workaround to the original problem you can check in loop wsrep_local_state_comment for existing tasks in entrypoint.sh and continue the task only when you find a node with wsrep_local_state_comment=Synced

not sure I understand correctly can you give an example ?  
when you start a node how would you know if it is the first node or an additional one ?


> And some protection from the race condition would be useful (that's why I rely on DISCOVERY SERVICE, as it provides atomic operations)

I really want to remove the dependancy for an external discovery service and I think I am very close.

**Comment by vadimtk** [comment]
@krasi-georgiev  it does not make a difference for the first node bootstrap.
What I mean if start additional nodes with --replicas=10 for example, the all will storm the single donor and some of them will fail.
So to avoid failure I propose checking if the donor is available by using `wsrep_local_state_comment=Synced`

However I am not sure how, without atomic operation, to resolve the case when multiple replicas will see that `wsrep_local_state_comment=Synced` and will try to join at the same time.

**Comment by krasi-georgiev** [comment]
@vadimtk yes that is a very good idea I will try it and do you have any idea how to force the very first node to bootstrap a new cluster ?

with dig -t A +short tasks.mysqlCluster
docker swarm will return only nodes reported healthy by the healthcheck so the loop can use that to return all available donors.

**Comment by krasi-georgiev** [comment]
@thaJeztah could you ping some of the other maintainers that can decide to reject or approve this proposal and I can work on implementing it.

**Comment by aluzzardi** [comment]
/cc @aaronlehmann @dongluochen 

Maybe we could use update parameters for initial rollout as well?

**Comment by aaronlehmann** [comment]
> Maybe we could use update parameters for initial rollout as well?

I'm fine with this in concept, but it would be a pretty big behavioral change. Maybe we need a flag to control the behavior?

**Comment by dongluochen** [comment]
I think treating initial deployment as rollout makes sense for this scenario. The default is to blast out all tasks like today. Flag (`--rollout`?) would allow initial deployment take the rollout model. It'd need changes to orchestrator/updater. 

I also hear request to allow user specified rollout sequence by node_id list, both in update and initial deployment. I'm not sure if this is a common use case.

**Comment by krasi-georgiev** [comment]
I think the concept is quite similiar to the 
`docker service update --update-parallelism=1` 
so maybe
`docker service create --parallelism=1`

**Comment by krasi-georgiev** [comment]
@thaJeztah has there been any decision on this? 
I would probably have some time to implement it.

**Comment by sirlatrom** [comment]
An (annoying) workaround could be to do `docker service create --name mysqlCluster --replicas 1 <some-image-with-healthcheck>` followed by `docker service update --replicas 3 --update-parallelism 1 --update-order start-first mysqlCluster`.

**Comment by krasi-georgiev** [comment]
I don't see any progress here so closing to cleanup my profile issues list.
